File: Hard/3336. Find the Number of Subsequences With Equal GCD/solution.py
class Solution:
    def subsequencePairCount(self, nums: List[int]) -> int:
        N = len(nums)

        dp = [[[-1 for k in range (201)]for j in range (201)]for i in range(N)]
        
        result = self.solve(0, 0, 0, nums, dp)

        return result

    def solve(self, idx: int, left: int, right: int, nums: List[int], dp: List[List[List[int]]]) -> int:
        if idx == len(nums):
            if left != 0 and right != 0 and left == right:
                return 1
            return 0

        if dp[idx][left][right] != -1:
            return dp[idx][left][right]

        MOD = 1000000007
        ret = 0

        ret = ret + self.solve(idx + 1, left, right, nums, dp)
        ret = ret % MOD

        ret = ret + self.solve(idx + 1, gcd(left, nums[idx]), right, nums, dp)
        ret = ret % MOD

        ret = ret + self.solve(idx + 1, left, gcd(right, nums[idx]), nums, dp)
        ret = ret % MOD

        dp[idx][left][right] = ret

        return dp[idx][left][right]

File: Hard/3336. Find the Number of Subsequences With Equal GCD/solution.cpp
class Solution {
public:
    int subsequencePairCount(vector<int>& nums) {
        int N = nums.size();

        vector<vector<vector<int>>> dp(
            N, vector<vector<int>> (
                201, vector<int>(201, -1)
            )
        );

        int result = solve(0, 0, 0, nums, dp);

        return result; 
    }

    int gcd(int a, int b) {
        if (a == 0) return b;
        if (b == 0) return a;
        if (a == b) return a;
        if (a > b) return gcd(a - b, b);
        return gcd(a, b - a);
    }

    int solve(int idx, int left, int right, vector<int>& nums, vector<vector<vector<int>>> &dp){
        if (idx == nums.size()){
            if (left != 0 && right != 0 && left == right)   return 1;
            return 0;
        }

        if (dp[idx][left][right] != -1) return dp[idx][left][right];

        int MOD = 1e9 + 7;
        long long ret = 0;

        ret += solve(idx + 1, left, right, nums, dp);
        ret %= MOD;

        ret += solve(idx + 1, gcd(left, nums[idx]), right, nums, dp);
        ret %= MOD;

        ret += solve(idx + 1, left, gcd(right, nums[idx]), nums, dp);
        ret %= MOD;

        return dp[idx][left][right] = ret;
    }
};